crates/bitwarden-crypto/src/store/context.rs

```diff
@@ -0,0 +1,531 @@
+// encryption/decryption. We also have the option to create a read/write context, which allows us to
+// modify the global keys, but only allows one context at a time. This is controlled by a RWLock on
+// the global keys, and this struct stores both types of guards.
+pub(crate) enum GlobalKeys<'a, Ids: KeyIds> {
+    ReadOnly(RwLockReadGuard<'a, KeyStoreInner<Ids>>),
+    ReadWrite(RwLockWriteGuard<'a, KeyStoreInner<Ids>>),
+}
+
+impl<Ids: KeyIds> GlobalKeys<'_, Ids> {
+    pub fn get(&self) -> &KeyStoreInner<Ids> {
+        match self {
+            GlobalKeys::ReadOnly(keys) => keys,
+            GlobalKeys::ReadWrite(keys) => keys,
+        }
+    }
+
+    pub fn get_mut(&mut self) -> Result<&mut KeyStoreInner<Ids>> {
+        match self {
+            GlobalKeys::ReadOnly(_) => Err(CryptoError::ReadOnlyKeyStore),
+            GlobalKeys::ReadWrite(keys) => Ok(keys),
+        }
+    }
+}
+
+impl<Ids: KeyIds> KeyStoreContext<'_, Ids> {
+    /// Clears all the local keys stored in this context
+    /// If this context has write access to the global keys,
+    /// it will also clear them
+    pub fn clear(&mut self) {
+        // Clear global keys if we have write access
+        if let Ok(keys) = self.global_keys.get_mut() {
+            keys.symmetric_keys.clear();
+            keys.asymmetric_keys.clear();
+        }
+
+        self.local_symmetric_keys.clear();
+        self.local_asymmetric_keys.clear();
+    }
+
+    /// Remove all symmetric keys from the context for which the predicate returns false
+    /// This will also remove the keys from the global store if this context has write access
+    pub fn retain_symmetric_keys(&mut self, f: fn(Ids::Symmetric) -> bool) {
+        if let Ok(keys) = self.global_keys.get_mut() {
+            keys.symmetric_keys.retain(f);
+        }
+        self.local_symmetric_keys.retain(f);
+    }
+
+    /// Remove all asymmetric keys from the context for which the predicate returns false
+    /// This will also remove the keys from the global store if this context has write access
+    pub fn retain_asymmetric_keys(&mut self, f: fn(Ids::Asymmetric) -> bool) {
+        if let Ok(keys) = self.global_keys.get_mut() {
+            keys.asymmetric_keys.retain(f);
+        }
+        self.local_asymmetric_keys.retain(f);
+    }
+
+    // TODO: All these encrypt x key with x key look like they need to be made generic,
+    // but I haven't found the best way to do that yet.
+
+    /// Decrypt a symmetric key into the context by using an already existing symmetric key
+    ///
+    /// # Arguments
+    ///
+    /// * `encryption_key` - The key id used to decrypt the `encrypted_key`. It must already exist
+    ///   in the context
+    /// * `new_key_id` - The key id where the decrypted key will be stored. If it already exists, it
+    ///   will be overwritten
+    /// * `encrypted_key` - The key to decrypt
+    pub fn decrypt_symmetric_key_with_symmetric_key(
```

Comment:
I wonder if the naming here is slightly misleading. We're not decrypting it we're adding it to the context?

Reply:
Well, we're doing both: We decrypt the key and add it to the context. Originally i was planning to make it a generic function like `decrypt_key_into_store`, but I couldn't get the typing to do what I wanted, which explains the long names.